kernel/src/capabilities.rs

```diff
@@ -14,71 +14,87 @@
 //! Capabilities are passed to modules from trusted code (i.e. code that can
 //! call `unsafe`).
 //!
-//! Capabilities are expressed as `unsafe` traits. Only code that can use
-//! `unsafe` mechanisms can instantiate an object that provides an `unsafe`
-//! trait. Functions that require certain capabilities require that they are
-//! passed an object that provides the correct capability trait. The object
-//! itself does not have to be marked `unsafe`.
+//! Capabilities are expressed as types wrapped in the [`Capability`] type.
+//! This wrapper cannot be constructed directly, only through the `unsafe`
+//! `new` method.
 //!
 //! Creating an object that expresses a capability is straightforward:
 //!
 //! ```
-//! use kernel::capabilities::ProcessManagementCapability;
+//! use kernel::capabilities::{Capability, ProcessManagement};
 //!
-//! struct ProcessMgmtCap;
-//! unsafe impl ProcessManagementCapability for ProcessMgmtCap {}
+//! let process_mgmt_cap = unsafe { Capability::<ProcessManagement>::new() };
 //! ```
 //!
-//! Now anything that has a ProcessMgmtCap can call any function that requires
-//! the `ProcessManagementCapability` capability.
+//! Now anything that has `process_mgmt_cap` can call any function that requires
+//! the `Capability<ProcessManagement>` capability.
 //!
 //! Requiring a certain capability is also straightforward:
 //!
 //! ```ignore
-//! pub fn manage_process<C: ProcessManagementCapability>(_c: &C) {
+//! pub fn manage_process(_c: &Capability<ProcessManagement>) {
 //!    unsafe {
-//!        ...
+//!        // ...
 //!    }
 //! }
 //! ```
 //!
 //! Anything that calls `manage_process` must have a reference to some object
-//! that provides the `ProcessManagementCapability` trait, which proves that it
-//! has the correct capability.
+//! of type `Capability<ProcessManagement>`, which proves that it has the
+//! correct capability.
+
+use core::marker::PhantomData;
+
+/// Wrapper type which can only be constructed through the unsafe `new` method.
+///
+/// Prevents capabilities from being created in safe code.
+pub struct Capability<T> {
+    /// Private type to prevent `Capability` being constructed manually.
+    capability: PhantomData<T>,
+}
+
+impl<T> Capability<T> {
+    /// Only allow `Capability`s to be created from `unsafe` code.
+    pub unsafe fn new() -> Self {
+        Self {
+            capability: PhantomData {},
+        }
+    }
+}
 
 /// The `ProcessManagementCapability` allows the holder to control
```

Comment:
Should you change the documentation as well?
`ProcessManagementCapability -> ProcessManagement`

Reply:
Yes, good spot.

I've done a grep for all the `{struct}Capability` strings and replaced them in comments. The only remaining matches come from the changelog and working group notes.